client/ts/src/channel/retriever.ts

```diff
@@ -128,27 +128,74 @@ export class CacheRetriever implements Retriever {
     const results: Payload[] = [];
     const toFetch: KeysOrNames = [];
     normalized.forEach((keyOrName) => {
-      const c = this.getFromCache(keyOrName);
-      if (c != null) results.push(c);
+      const c = this.get(keyOrName);
+      if (c != null) results.push(...c);
       else toFetch.push(keyOrName as never);
     });
     if (toFetch.length === 0) return results;
     const fetched = await this.wrapped.retrieve(toFetch, options);
-    this.updateCache(fetched);
+    this.set(fetched);
     return results.concat(fetched);
   }
 
-  private updateCache(channels: Payload[]): void {
+  delete(keys: Params): void {
+    const { variant, normalized } = analyzeChannelParams(keys);
+    if (variant === "names")
+      (normalized as string[]).forEach((name) => {
+        const keys = this.namesToKeys.get(name);
+        if (keys == null) return;
+        keys.forEach((k) => this.cache.delete(k));
+        this.namesToKeys.delete(name);
+      });
+    else
+      (normalized as number[]).forEach((key) => {
+        const channel = this.cache.get(key);
+        if (channel == null) return;
+        this.cache.delete(key);
+        this.namesToKeys.delete(channel.name);
+      });
+  }
+
+  rename(keys: Key[], names: string[]): void {
+    keys.forEach((key, i) => {
+      const name = names[i];
+      const ch = this.cache.get(key);
+      if (ch == null) return;
+      this.cache.delete(key);
+      const keys = this.namesToKeys.get(ch.name);
+      if (keys != null) keys.delete(key);
```

Comment:
This results in empty sets in the cache – I would add a conditional deleting the entire set if there is only one element of that name (this is also the hot path)

Reply:
Will do